pallets/domains/src/lib.rs

```
@@ -0,0 +1,517 @@
+                Error::<T>::TooManyDomainsPerAccount,
+            );
+
+            let expires_at = expires_in.saturating_add(System::<T>::block_number());
+
+            let deposit = T::BaseDomainDeposit::get();
+            let domain_meta = DomainMeta::new(
+                full_domain.clone(),
+                owner.clone(),
+                content,
+                expires_at,
+                deposit,
+            );
+
+            if let ReserveDeposit::Yes = reserve_deposit {
+                <T as Config>::Currency::reserve(&owner, deposit)?;
+            }
+
+            // TODO: withdraw balance
+
+            RegisteredDomains::<T>::insert(domain_lc.clone(), domain_meta);
+            DomainsByOwner::<T>::mutate(
+                &owner, |domains| {
+                    domains.try_push(domain_lc.clone()).expect("qed; too many domains per account")
+                }
+            );
+
+            Self::deposit_event(Event::DomainRegistered(owner, full_domain));
+            Ok(())
+        }
+
+        /// Throws an error if domain contains invalid character.
+        fn ensure_domain_contains_valid_chars(domain: &[u8], error: Error<T>) -> DispatchResult {
+            let is_char_alphanumeric = |c: &&u8| (**c).is_ascii_alphanumeric();
+
+            let first_char_alphanumeric = domain.first().filter(is_char_alphanumeric).is_some();
+            let last_char_alphanumeric = domain.last().filter(is_char_alphanumeric).is_some();
+
+            let mut prev_char_hyphen = false;
+            let domain_correct = domain.iter().all(|c| {
+                let curr_char_hyphen = *c == b'-';
+
+                if prev_char_hyphen && curr_char_hyphen {
+                    return false;
+                }
+
+                prev_char_hyphen = curr_char_hyphen;
+                c.is_ascii_alphanumeric() || curr_char_hyphen
+            });
+
+            ensure!(first_char_alphanumeric && last_char_alphanumeric && domain_correct, error);
```

You can fail faster even before the loop that checks for valid characters with:
`ensure!(first_char_alphanumeric && last_char_alphanumeric, error)`